### PR TITLE
feat: add crossvalidator to v2

### PR DESCRIPTION
**Pull request by sarakolding**
# Notes for reviewers
we will still run into issues when implementing multilabel (since we hard code the column name "y_hat_probs"), but that's a general issue, so maybe let it be as is for now?

### CODE REVIEW

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 1]
```diff
@@ -1,30 +1,105 @@
...
 
 
+@BaselineRegistry.trainers.register("crossval_trainer")
 class CrossValidatorTrainer(BaselineTrainer):
     def __init__(
         self,
-        data: PolarsFrame,
-        outcome_col_name: str,
+        training_data: BaselineDataLoader,
+        training_outcome_col_name: str,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:24]:
```suggestion
        outcome_col_name: str,
```

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 2]
```diff
@@ -1,30 +1,105 @@
...
         task: BaselineTask,
-        n_splits: int,
+        metric: BaseMetric,
         logger: BaselineLogger,
+        n_splits: int = 5,
+        group_col_name: str = "dw_ek_borger",
     ):
-        ...
+        self.training_data = training_data.load()
+        self.training_outcome_col_name = training_outcome_col_name
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:33]:
```suggestion
        self.outcome_col_name = outcome_col_name
```

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 3]
```diff
@@ -1,30 +1,105 @@
...
+        self.logger = logger
 
     def train(self) -> TrainingResult:
-        ...
+        training_data_preprocessed = self.preprocessing_pipeline.apply(
+            data=self.training_data,
+        )
+
+        X = training_data_preprocessed.drop(
+            self.training_outcome_col_name,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:47]: This assumes that only one outcome column is in the dataset. We probably want to protect ourselves against that assumption. E.g. if we run this with my T2D dataset, 4 of the outcome columns will be in the training data.

That's a step for preprocessing, though, not this PR. I'll create an issue.

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 4]
```diff
@@ -1,30 +1,105 @@
...
+            columns=[self.training_outcome_col_name],
+        )
+
+        folds = StratifiedGroupKFold(n_splits=self.n_splits).split(
+            X=X,
+            y=y,
+            groups=training_data_preprocessed[self.group_col_name],
+        )
+
+        for _i, (train_idxs, val_idxs) in enumerate(folds):
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:61]:
```suggestion
        for _, (train_idxs, val_idxs) in enumerate(folds):
```

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 5]
```diff
@@ -1,30 +1,105 @@
...
+            self.task.train(X_train, y_train, y_col_name=self.training_outcome_col_name)
+
+            y_hat_prob = self.task.predict_proba(
+                X_train,
+            )
+
+            self.logger.log_metric(
+                self.metric.calculate(
+                    y_train[self.training_outcome_col_name],
+                    y_hat_prob,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:76]:
```suggestion
                    y=y_train[self.training_outcome_col_name],
                    y_hat_probs=y_hat_prob,
```

Comment (MartinBernstorff) [reply psycop/common/model_training_v2/trainer/cross_validator.py:76]: To ensure we don't mix them up :-)

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 6]
```diff
@@ -1,30 +1,105 @@
...
+            )
+
+            oof_y_hat_prob = self.task.predict_proba(
+                X.loc[val_idxs],
+            )
+
+            self.logger.log_metric(
+                self.metric.calculate(
+                    y.loc[val_idxs][self.training_outcome_col_name],
+                    oof_y_hat_prob,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:87]: Same here.

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 7]
```diff
@@ -1,30 +1,105 @@
...
+                    y_train[self.training_outcome_col_name],
+                    y_hat_prob,
+                ),
+            )
+
+            oof_y_hat_prob = self.task.predict_proba(
+                X.loc[val_idxs],
+            )
+
+            self.logger.log_metric(
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:84]: We probably want to run `logger.info` here, since this is more inspectability than the final metric :-)

File: psycop/common/model_training_v2/trainer/cross_validator.py [hunk 8]
```diff
@@ -1,30 +1,105 @@
...
+                y.loc[train_idxs],
+            )
+
+            self.task.train(X_train, y_train, y_col_name=self.training_outcome_col_name)
+
+            y_hat_prob = self.task.predict_proba(
+                X_train,
+            )
+
+            self.logger.log_metric(
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/cross_validator.py:73]: We probably want to run `logger.info` here, since this is more inspectability than the final metric :-)

File: psycop/common/model_training_v2/trainer/cross_validator.py (hunk 7 above)
Comment (MartinBernstorff) [reply psycop/common/model_training_v2/trainer/cross_validator.py:84]: Alternatively, we should rename the metric so it's logged under 'oof_{fold_i}_{metric_name}' or similar.

**Review by MartinBernstorff — CHANGES_REQUESTED** [review]

File: psycop/common/model_training_v2/trainer/cross_validator.py (hunk 8 above)
Comment (sarakolding) [reply psycop/common/model_training_v2/trainer/cross_validator.py:73]: rename every i fold

**Review by MartinBernstorff — APPROVED** [review]
